File: android/app/src/main/java/com/albunyaan/tube/util/EmailShape.kt
package com.albunyaan.tube.util

/**
 * RFC-5322-shaped email validator. Mirrors Firebase Auth's own minimum
 * shape: a single non-empty local part, one '@', and a domain with at
 * least one dot (no leading or trailing dot). Pre-network gate so we
 * don't burn Firebase throttle quota on obviously-malformed input.
 */
fun isEmailShape(s: String): Boolean {
    val at = s.indexOf('@')
    if (at <= 0 || at != s.lastIndexOf('@')) return false
    if (at == s.length - 1) return false
    val domain = s.substring(at + 1)
    return domain.contains('.') && !domain.startsWith('.') && !domain.endsWith('.')
}

File: ios/FitrahTube/Features/Auth/EmailShape.swift
import Foundation

/// The pre-network address gate (`EmailShape.kt:9-15`, ported verbatim). Deliberately NOT an
/// RFC-5322 validator: it mirrors Firebase Auth's own minimum shape, so it can never reject an
/// address Firebase would have accepted — it only stops the ones Firebase would immediately reject,
/// which is what keeps malformed attempts from burning the IP-based throttle window that legitimate
/// users on flaky networks then hit.
///
/// One `@`, a non-empty local part, and a domain carrying a dot that neither opens nor closes it.
nonisolated enum EmailShape {
    static func isValid(_ address: String) -> Bool {
        guard let at = address.firstIndex(of: "@"),
              at != address.startIndex,                       // non-empty local part
              address.lastIndex(of: "@") == at,               // exactly one "@"
              at != address.index(before: address.endIndex)   // non-empty domain
        else { return false }
        let domain = address[address.index(after: at)...]
        return domain.contains(".") && !domain.hasPrefix(".") && !domain.hasSuffix(".")
    }
}
